design/secretless-e2e-tests-dnd.md

```diff
@@ -0,0 +1,276 @@
+    ```
+    [
+      {
+          "config": "./mysql_config.json",
+          "service_definition": "mysql_service",
+      },
+      {
+          "config": "./conjur_config.json",
+          "service_definition": "conjur_service"
+      }
+    ]
+    ```
+1. There is a test case written in Gherkin for validating using the MySQL service connector with Conjur. See below that there is an expliit step specifying the services that are expected to be running. 
+    ```
+    Feature: Secretless happy path with database connectors
+    
+      Scenario: Connection to MySQL with MySQL credentials stored in Conjur
+        Given services are running as defined in "./services.json" and the results are store in context path "services"
+        When I store secrets defined in "./conjur-secrets.json" to Conjur using credentials in context path "services.conjur_service.creds"
+        And I create an identity for an application as defined in "./application_identity.json" using credentials in context path "services.conjur_service.creds""
+        And I deploy a test app as defined in "./test-app.yml"
+        Then the app is able to connect to MYSQL on "localhost:332"
+    ```
+1. The command `cucumber` is called to run test cases. The command is configurable and allows the runner to select a subset of all tests to run.
+
+## Interfaces
+
+### Opaque API for lifecycle management of dependency resources
+
+The API makes it possible to implement service lifecycle methods in whatever way makes sense. That could be literal bash, docker-compose, calling an HTTP API to grab the service from a pool etc. This makes it possible to have similar experience between different environments (Jenkins, developer machine etc.). The same exact commands would be called on Jenkins as on the developer machine. The action of the commands would depend on the environment. For example, the developer machine can use a less resource intensive implementation of the service and Jenkins can run something that more closely approximates customer environments.
+
+### Steps towards implementation
+
+1. In the short term we can change `kubernetes-conjur-demo` so that the Service Definition for Conjur masks calls to `kubernetes-conjur-deploy`
+2. Later the infrastructure team or whoever can fork and modify `kubernetes-conjur-deploy` (or start from scratch). This means improvements (e.g. optimise startup time, performance, customer-environment accuracy continue) could be made to the service implementation without affecting existing tests. As long as the interface contract is enforced all things remain equal.
+
+## Testing
+
+TBD
+
+## Infrastructure
+
+The orchestration of dependencies (fixtures, services etc.) will need to be run somewhere. Some dependencies such as Conjur followers need to be run on the same infrastructure as the tests. Some dependencies such as target services can and perhaps should be run on separate infrastructure from the tests.
```

Comment:
Running followers means that you also run somehow a master or a master cluster, somewhere outside of this environment. So we'll need to decide if a master is assumed to be continuously available and running, or if it's provisioned as well.